src/blueprint.zig

```
@@ -0,0 +1,330 @@
+	}
+	fn getBlockArraySizeBytes(self: @This()) usize {
+		return self.getBlockArraySize()*@sizeOf(BlockStorageType);
+	}
+	pub fn getHeaderSizeBytes(_: @This()) usize {
+		var size: usize = 0;
+		inline for(@typeInfo(@This()).@"struct".fields) |field| {
+			size += @sizeOf(field.type);
+		}
+		return size;
+	}
+	pub fn getDecompressedDataSizeBytes(self: @This()) usize {
+		return self.paletteSizeBytes + self.getBlockArraySizeBytes();
+	}
+};
+
+pub const Blueprint = struct {
+	blocks: main.List(Block),
+	sizeX: usize,
+	sizeY: usize,
+	sizeZ: usize,
+
+	pub fn init(allocator: NeverFailingAllocator) @This() {
+		return Blueprint{
+			.blocks = .init(allocator),
+			.sizeX = 0,
+			.sizeY = 0,
+			.sizeZ = 0,
+		};
+	}
+	pub fn deinit(self: *@This()) void {
+		self.blocks.deinit();
+	}
+	pub fn clear(self: *@This()) void {
+		self.sizeX = 0;
+		self.sizeY = 0;
+		self.sizeZ = 0;
+		self.blocks.clearRetainingCapacity();
+	}
+	pub fn capture(self: *@This(), pos1: Vec3i, pos2: Vec3i) void {
+		self.clear();
+
+		const startX = @min(pos1[0], pos2[0]);
+		const startY = @min(pos1[1], pos2[1]);
+		const startZ = @min(pos1[2], pos2[2]);
+
+		const endX = @max(pos1[0], pos2[0]);
+		const endY = @max(pos1[1], pos2[1]);
+		const endZ = @max(pos1[2], pos2[2]);
+
+		const sizeX: usize = @intCast(@abs(endX - startX + 1));
+		const sizeY: usize = @intCast(@abs(endY - startY + 1));
+		const sizeZ: usize = @intCast(@abs(endZ - startZ + 1));
+
+		self.sizeX = sizeX;
+		self.sizeY = sizeY;
+		self.sizeZ = sizeZ;
+
+		for(0..sizeX) |offsetX| {
+			const worldX = startX + @as(i32, @intCast(offsetX));
+
+			for(0..sizeY) |offsetY| {
+				const worldY = startY + @as(i32, @intCast(offsetY));
+
+				for(0..sizeZ) |offsetZ| {
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const block = main.server.world.?.getBlock(worldX, worldY, worldZ) orelse Block{.typ = 0, .data = 0};
```

If the chunk is not loaded, then we should return an error and message the user in the command feedback instead of silently failing.